Annotation with brushing tools (#5269)
opencv/cvat: site/content/en/docs/manual/advanced/annotation-with-brushing-tools.md

---
title: 'Annotation with brushing tools'
linkTitle: 'Annotation with brushing tools'
weight: 13
description: 'Guide to annotating tasks using  brushing tools.'
---

> The brushing tool creates a segmentation mask (further - mask) on the top of the image. Use it, when you need to add a label to an object that has several separated parts, but must be counted as one, for example - a house hiding behind trees or a pillar behind a traffic sign.

Reviewer: actually, not on the top of the image

Reviewer 2: I suggest to put it this way: 

```suggestion
The brush tool allows you to draw mask annotations. This type of annotations is mainly used in Semantic (Panoptic) and Instance Image Segmentation tasks. You can read more about working with segmentation masks in CVAT [here](/docs/manual/advanced/annotation-with-polygons/creating-mask/).

Comparing to polygons, this tool can be particularly useful to create masks for disjoint objects, having multiple parts, such as a house hiding behind trees, a car occluded by a pedestrian, or a pillar behind a traffic sign. In some cases, it can also be simpler and faster to draw a mask with a brush than a polygon. The brush tool has different modes that can be useful in different annotation scenarios. For instance, there are modes to erase pixels, change brush shapes and to draw polygon-like contours.
```

Author: By a number of style manuals, It is advised to avoid "allows"(advice from multiple manuals of styles), And  "comparing to polygons" sounds a bit confusing. The readers might know nothing about polygon tools and we do not want them to start thinking that he/she has to learn what is polygon first. Or do we?

Author: We also must avoid passive voice as much as possible (it is just a piece of general writing advice).

> You can annotate images manually or upload already annotated images, you can also convert polygon annotation to masks and the other way around.

Reviewer: > can also convert polygon annotation to masks

How is it possible? I only can see one way to do that, is to export dataset in one of segmentation masks format and import it back with disabled option 'Convert masks to polygons'. The tool itself does not provide such conversion capabilities on UI.

Reviewer 2: This paragraph seems extra here.

You can use annotation with brush together with polygon-to-mask annotation on the same image.

See:

- [Brushing tools menu](#brushing-tools-menu)
- [Annotation with brush](#annotation-with-brush)
- [Annotation with polygon-to-mask](#annotation-with-polygon-to-mask)
- [Remove underlying pixels](#remove-underlying-pixels)
- [Import and export](#import-and-export)

## Brushing tools menu

The brushing tool menu appears on the top of the screen after you click **Draw a mask**:

![BT Menu](/images/brushing_tool_menu.png)

It has the following elements:

|Element|Description|
|----|-----|
|![Tick icon](/images/tick_icon.png)| **Save mask** saves the created  mask. The saved mask will appear on the object sidebar|

> |![Add label](/images/brushing_tools_add_label.png)| **Add label** adds a new label to the object sidebar, brushing tool menu switches to **Brush**.|

Reviewer: ```suggestion
|![Add label](/images/brushing_tools_add_label.png)| **Save mask and continue** adds a new mask to the object sidebar, and allows you to draw a new one immediately.|
```

> |![Brush](/images/brushing_tools_icon.png)| **Brush** applies the mask to the object on the image.|

Reviewer: ```suggestion
|![Brush](/images/brushing_tools_icon.png)| **Brush** used to add new image regions to a mask.|
```

> |![Eraser](/images/brushing_tools_erase.png)|**Eraser** removes part of the mask from the image.|

Reviewer: ```suggestion
|![Eraser](/images/brushing_tools_erase.png)|**Eraser** removes part of the images from the mask.|
```

Reviewer 2: ```suggestion
|![Eraser](/images/brushing_tools_erase.png)| **Eraser** allows to remove parts of the mask.|
```

> |![Add poly](/images/brushing_tools_add_poly.png)|**Polygon**  selection tool. Selection will be converted to the mask.|

Reviewer: ```suggestion
|![Add poly](/images/brushing_tools_add_poly.png)|**Polygon**  selection tool. Selection will be added to the drawn mask.|
```

> |![Remove poly](/images/brushing_tools_remove_poly.png)|**Remove polygon selection** removes a part of the polygon selection.|

Reviewer: ```suggestion
|![Remove poly](/images/brushing_tools_remove_poly.png)|**Remove polygon selection**. Selection will be substracted from the drawn mask.|
```

Author: done

|![Brush size](/images/brushing_tools_brush_size.png)|**Brush size** in pixels. <br>**Note:** Visible only when **Brush** is selected.|

> |![Brush shape](/images/brushing_tools_brush_shape.png)|**Brush shape** with two options: circle and square. <br>**Note:** Visible only when **Brush** is selected.|

Reviewer: Visible when one of ``brush`` or ``eraser`` are choosen.

|![Pixel remove](/images/brushing_tools_pixels.png)|**Remove underlying pixels**. When you are drawing or editing a mask with this tool, <br>pixels on other masks that are located at the same positions as the pixels of the <br>current mask are deleted.|

> |![Label](/images/brushing_tools_label_drop.png)|**Label** that will be assigned to the  mask|

Reviewer: ```suggestion
|![Label](/images/brushing_tools_label_drop.png)|**Label** that will be assigned to the newly created mask|
```

|![Move](/images/brushing_tools_brush_move.png)|**Move**. Click and hold to move the menu bar to the other place on the screen|

## Annotation with brush

Prerequisites: [Create task](/docs/manual/basics/creating_an_annotation_task/), upload the image, add labels, and open the workspace.

Do the following:

1. From the [controls sidebar](/docs/manual/basics/controls-sidebar/), select **Brush** ![Brush icon](/images/brushing_tools_icon.png).
2. In the **Draw new mask** menu, select label for your mask, and click **Draw a mask**. <br>The **Brush**![Brush](/images/brushing_tools_icon.png) tool will be selected by default.

   ![BT context menu](/images/brushing_tools_context_menu.png)

> 3. With the brush, draw a mask on the object you want to label. <br>To correct selection, use **Eraser** ![Eraser](/images/brushing_tools_erase.png)

Reviewer: 1. ``correct`` assumes not only removing regions from the mask, so it is not about **eraser**
2. not only eraser, but also polygon-minus tool is possible to remove a part of mask

I believe here we should just refer a user to the tools overview above.

Author: changed to "erase" for the polygon updated in the different section

>    ![Brushing](/images/brushing_tools.gif)

Reviewer 3: Just in my opinion, I would suggest to use more contrast colors to demonstrate the tool usage. Here we see brown on brown and it's difficult to recognize the annotation result.

Author: updated

4. After you applied the mask, on the top menu bar click **Save mask** ![Tick icon](/images/tick_icon.png) to finish the process (or **N** on the keyboard).
5. Added label will appear on the [objects sidebar](/docs/manual/basics/objects-sidebar/).

To add the next label, repeat steps 1 to 5. All added labels will be visible on the image and the  [objects sidebar](/docs/manual/basics/objects-sidebar/).

> To save the project with all added labels, on the top menu click **Save** ![Save](/images/brushing_tools_save.png).

Reviewer: ```suggestion
To save the job with all added objects, on the top menu click **Save** ![Save](/images/brushing_tools_save.png).
```

   ![BT context menu](/images/brushing_tools_context_menu.png)

3. In the brushing tool menu, select **Polygon** ![Add poly](/images/brushing_tools_add_poly.png).
4. With the **Polygon**![Add poly](/images/brushing_tools_add_poly.png) tool, draw a mask for the object you want to label. <br>To correct selection, use **Remove polygon selection** ![Remove poly](/images/brushing_tools_remove_poly.png).
5. Use **Save mask** ![Tick icon](/images/tick_icon.png) (or **N** on the keyboard) to switch between add/remove polygon tools:

   ![Brushing](/images/brushing_tools_polygon.gif)

6. After you added the polygon selection, on the top menu bar click **Save mask** ![Tick icon](/images/tick_icon.png) to finish the process (or **N** on the keyboard).
7. Added label will appear on the [objects sidebar](/docs/manual/basics/objects-sidebar/).

To add the next label, repeat steps 1 to 5. All added labels will be visible on the image and the  [objects sidebar](/docs/manual/basics/objects-sidebar/).

To save the project with all added labels, on the top menu click **Save** ![Save](/images/brushing_tools_save.png).

## Remove underlying pixels

> Use **Remove underlying pixels** tool when you want to draw a new mask on top of the other masks and simultaneously delete the pixels of other masks that are located at the same positions.

Reviewer: ```suggestion
Use **Remove underlying pixels** tool when you want to delete the pixels of other masks that are located at the same positions. It is highly useful feature to avoid meticulous drawing edge twice between two different objects
```
